lib/gvariant.nix

```diff
@@ -0,0 +1,181 @@
+# This file is taken from https://github.com/nix-community/home-manager
+# Copyright (c) 2017-2022 Home Manager contributors
+#
+# A partial and basic implementation of GVariant formatted strings.
+#
+# Note, this API is not considered fully stable and it might therefore
+# change in backwards incompatible ways without prior notice.
+
+{ lib }:
+
+let
+  inherit (lib)
+    concatMapStringsSep concatStrings escape hasPrefix head replaceStrings;
+
+  mkPrimitive = t: v: {
+    _type = "gvariant";
+    type = t;
+    value = v;
+    __toString = self: "@${self.type} ${toString self.value}";
+  };
+
+  type = {
+    arrayOf = t: "a${t}";
+    maybeOf = t: "m${t}";
+    tupleOf = ts: "(${concatStrings ts})";
+    dictionaryEntryOf = ts: "{${concatStrings ts}}";
+    string = "s";
+    boolean = "b";
+    uchar = "y";
+    int16 = "n";
+    uint16 = "q";
+    int32 = "i";
+    uint32 = "u";
+    int64 = "x";
+    uint64 = "t";
+    double = "d";
+    variant = "v";
+  };
+
+  # Returns the GVariant type of a given Nix value. If no type can be
+  # found for the value then the empty string is returned.
+  typeOf = v:
+    with type;
+    if builtins.isBool v then
+      boolean
+    else if builtins.isInt v then
+      int32
+    else if builtins.isFloat v then
+      double
+    else if builtins.isString v then
+      string
+    else if builtins.isList v then
+      let elemType = elemTypeOf v;
+      in if elemType == "" then "" else arrayOf elemType
+    else if builtins.isAttrs v && v ? type then
+      v.type
+    else
+      "";
+
+  elemTypeOf = vs:
+    if builtins.isList vs then
+      if vs == [ ] then "" else typeOf (head vs)
+    else
+      "";
+
+  mkMaybe = elemType: elem:
+    mkPrimitive (type.maybeOf elemType) elem // {
+      __toString = self:
+        if self.value == null then
+          "@${self.type} nothing"
+        else
+          "just ${toString self.value}";
+    };
+
+in
+rec {
+
+  inherit type typeOf;
+
+  isGVariant = v: v._type or "" == "gvariant";
+
+  isArray = hasPrefix "a";
+  isDictionaryEntry = hasPrefix "{";
+  isMaybe = hasPrefix "m";
+  isTuple = hasPrefix "(";
+
+  # Returns the GVariant value that most closely matches the given Nix
+  # value. If no GVariant value can be found then `null` is returned.
+
+  mkValue = v:
```

Comment:
I added docs for every functions.

Reply:
Nice. Please also add it to https://github.com/NixOS/nixpkgs/blob/2ba28b9c95690e40187c99d6756e98ac58d678b2/doc/doc-support/default.nix#L6 so that it is included in the Nixpkgs manual (buildable with `nix-build pkgs/top-level/release.nix -A manual`).